### PR TITLE
Introduce CAA worker

### PR DESCRIPTION
**Pull request by 0xdiba**
The worker checks the domain's DNS records for CAA and reports CAs permitted to issue standard and wildcard certs ( if any ).

### CODE REVIEW

File: worker/caaWorker/caaWorker.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+	hostPieces := strings.Split(in.Target, ".")
+	for i := 0; i < len(hostPieces); i++ {
+		host := strings.Join(hostPieces[i:], ".")
+
+		msg := new(dns.Msg)
+		msg.SetQuestion(dns.Fqdn(host), dns.TypeCAA)
+
+		client := dns.Client{}
+		res, _, err := client.Exchange(msg, "8.8.8.8:53")
```
Comment (jvehent) [inline worker/caaWorker/caaWorker.go]: Should this use the locally configured resolver?

File: worker/caaWorker/caaWorker.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+func init() {
+	worker.RegisterWorker(workerName, worker.Info{Runner: new(eval), Description: workerDesc})
+}
+
+// Result describes the result produced by CAAWorker
+type Result struct {
+	HasCAA       bool     `json:"has_caa"`
+	Host         string   `json:"host"`
+	IssueCAs     []string `json:"issue_cas"`
+	IssueWildCAs []string `json:"issuewild_cas"`
```
Comment (jvehent) [inline worker/caaWorker/caaWorker.go]: I think the `_cas` suffix to these json tags could be removed.

File: worker/caaWorker/caaWorker.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+	} else {
+		result.Success = true
+		result.Result = res
+	}
+
+	resChan <- result
+}
+
+// Assertor compares 2 caaResults and reports differences.
+func (e eval) Assertor(caaResult, assertresults []byte) (pass bool, body []byte, err error) {
```
Comment (jvehent) [inline worker/caaWorker/caaWorker.go]: `tls-observatory.services.mozilla.com` has caa under `services.mozilla.com`. How about adding that to runner.yaml as an example of using this assertor?

File: worker/caaWorker/caaWorker.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+func (e eval) AnalysisPrinter(input []byte, printAll interface{}) (results []string, err error) {
+	var r Result
+	err = json.Unmarshal(input, &r)
+	if err != nil {
+		err = fmt.Errorf("CAA worker: failed to parse results: %v", err)
+		return
+	}
+
+	if !r.HasCAA {
+		results = append(results, "No avaliable CAA records")
```
Comment (jvehent) [inline worker/caaWorker/caaWorker.go]: typo: `available`

File: worker/caaWorker/caaWorker.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+	} else {
+		results = append(results, fmt.Sprintf("Available CAA records for domain: %s", r.Host))
```
Comment (jvehent) [inline worker/caaWorker/caaWorker.go]: A shorter string reads better in console output: `* CAA records: {found|not found}`
Also, use a `*` for worker name, then `-` for worker details.

File: worker/caaWorker/caaWorker.go [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+		if len(r.IssueCAs) > 0 {
+			results = append(results, "- CAs permitted to issue certificates:")
```
Comment (jvehent) [inline worker/caaWorker/caaWorker.go]: Add 2 white spaces before the hyphen: `  - CA $CA permitted to issue certs for $host`

File: worker/caaWorker/caaWorker.go [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+			for _, issue := range r.IssueCAs {
+				results = append(results, fmt.Sprintf("  * %s", issue))
+			}
+		}
+		if len(r.IssueWildCAs) > 0 {
+			results = append(results, "- CAs permitted to issue wildcard certificates:")
```
Comment (jvehent) [inline worker/caaWorker/caaWorker.go]: Add 2 white spaces before the hyphen: `  - CA $CA permitted to issue wildcard certs for $host`

File: worker/caaWorker/caaWorker.go [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+			for _, wild := range r.IssueWildCAs {
+				results = append(results, fmt.Sprintf("  * %s", wild))
+			}
+		}
+	}
```
Comment (jvehent) [inline worker/caaWorker/caaWorker.go]: Final output should read like this:
```
* CAA records: found
  - CA 'digicert.com' permitted to issue certs for 'services.mozilla.com'
```

**Comment by jvehent** [comment]
Need to import `caaWorker` into `tlsobs/main.go`
```diff
diff --git a/tlsobs/main.go b/tlsobs/main.go
index e7fe71d6..704a0856 100644
--- a/tlsobs/main.go
+++ b/tlsobs/main.go
@@ -17,6 +17,7 @@ import (
        "github.com/mozilla/tls-observatory/connection"
        "github.com/mozilla/tls-observatory/database"
        "github.com/mozilla/tls-observatory/worker"
+       _ "github.com/mozilla/tls-observatory/worker/caaWorker"
        _ "github.com/mozilla/tls-observatory/worker/mozillaEvaluationWorker"
        _ "github.com/mozilla/tls-observatory/worker/mozillaGradingWorker"
        _ "github.com/mozilla/tls-observatory/worker/sslLabsClientSupport"
```

Then it works:
```
--- Analyzers ---
Available CAA records for domain: services.mozilla.com
- CAs permitted to issue certificates:
  * digicert.com
```

File: worker/caaWorker/caaWorker.go (hunk 3 above)
Comment (0xdiba) [reply worker/caaWorker/caaWorker.go]: Done

File: worker/caaWorker/caaWorker.go (hunk 1 above)
Comment (0xdiba) [reply worker/caaWorker/caaWorker.go]: I am not sure about what the caching behaviour is and if CAA records and fetched correctly.

Comment (jvehent) [reply worker/caaWorker/caaWorker.go]: But people running tlsobs locally may not permit outbound DNS to Google, in which case this worker will fail. I think reading the resolver from /etc/resolv.conf and falling back to 8.8.8.8 is the right thing to do.

Comment (0xdiba) [reply worker/caaWorker/caaWorker.go]: Done, nice catch :+1:

**Pull request merged**